<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getBlastFromExcel</title>
  <meta name="keywords" content="getBlastFromExcel">
  <meta name="description" content="getBlastFromExcel">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">external</a> &gt; getBlastFromExcel.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getBlastFromExcel
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getBlastFromExcel</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function blastStructure=getBlastFromExcel(models,blastFile,organismId) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getBlastFromExcel
   Retrieves gene homology information from Excel files. Used as
   input to getModelFromHomology.

   models          a cell array of model structures
   blastFile       Excel file with homology information
   organismId      the id of the organism of interest (as described in the
                   Excel file)

   blastStructure  structure containing the information in the Excel
                   sheets.

   The Excel file should contain a number of spreadsheets which in turn
   contain the bidirectional homology measurements between the genes in the
   organisms. The first and second column headers in each sheet is the
   &quot;to&quot; and &quot;from&quot; model ids (as defined in models or for the new organism).
   The entries should correspond to the gene names in those models. The third,
   fourth, and fifth columns represent the E-value, alignment length, and
   identity for each measurement (captions should be &quot;E-value&quot;, &quot;Alignment length&quot;,
   and &quot;Identity&quot;).

   Usage: blastStructure=getBlastFromExcel(models,blastFile,organismId)

   Eduard Kerkhoven, 2018-05-18</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function blastStructure=getBlastFromExcel(models,blastFile,organismId)</a>
0002 <span class="comment">% getBlastFromExcel</span>
0003 <span class="comment">%   Retrieves gene homology information from Excel files. Used as</span>
0004 <span class="comment">%   input to getModelFromHomology.</span>
0005 <span class="comment">%</span>
0006 <span class="comment">%   models          a cell array of model structures</span>
0007 <span class="comment">%   blastFile       Excel file with homology information</span>
0008 <span class="comment">%   organismId      the id of the organism of interest (as described in the</span>
0009 <span class="comment">%                   Excel file)</span>
0010 <span class="comment">%</span>
0011 <span class="comment">%   blastStructure  structure containing the information in the Excel</span>
0012 <span class="comment">%                   sheets.</span>
0013 <span class="comment">%</span>
0014 <span class="comment">%   The Excel file should contain a number of spreadsheets which in turn</span>
0015 <span class="comment">%   contain the bidirectional homology measurements between the genes in the</span>
0016 <span class="comment">%   organisms. The first and second column headers in each sheet is the</span>
0017 <span class="comment">%   &quot;to&quot; and &quot;from&quot; model ids (as defined in models or for the new organism).</span>
0018 <span class="comment">%   The entries should correspond to the gene names in those models. The third,</span>
0019 <span class="comment">%   fourth, and fifth columns represent the E-value, alignment length, and</span>
0020 <span class="comment">%   identity for each measurement (captions should be &quot;E-value&quot;, &quot;Alignment length&quot;,</span>
0021 <span class="comment">%   and &quot;Identity&quot;).</span>
0022 <span class="comment">%</span>
0023 <span class="comment">%   Usage: blastStructure=getBlastFromExcel(models,blastFile,organismId)</span>
0024 <span class="comment">%</span>
0025 <span class="comment">%   Eduard Kerkhoven, 2018-05-18</span>
0026 <span class="comment">%</span>
0027 
0028 <span class="keyword">if</span> ~(exist(blastFile,<span class="string">'file'</span>)==2)
0029     error(<span class="string">'BLAST result file %s cannot be found'</span>,string(blastFile));
0030 <span class="keyword">end</span>
0031 
0032 blastStructure=[];
0033 
0034 <span class="comment">%Get a list of model IDs</span>
0035 organisms=cell(numel(models)+1,1);
0036 organisms{1}=organismId;
0037 <span class="keyword">for</span> i=1:numel(models)
0038     organisms{i+1}=models{i}.id;
0039 <span class="keyword">end</span>
0040 
0041 <span class="comment">%Get all the spreadsheets in the file</span>
0042 [type, sheets]=xlsfinfo(blastFile);
0043 
0044 <span class="comment">%Check if the file is a Microsoft Excel Spreadsheet</span>
0045 <span class="keyword">if</span> ~strcmp(type,<span class="string">'Microsoft Excel Spreadsheet'</span>)
0046     EM=<span class="string">'The file is not a Microsoft Excel Spreadsheet'</span>;
0047     dispEM(EM);
0048 <span class="keyword">end</span>
0049 
0050 <span class="keyword">for</span> i=1:numel(sheets)
0051     <span class="comment">%Check if the sheet has the right header and deal with organisms that</span>
0052     <span class="comment">%are in &quot;models&quot;</span>
0053     [values,dataSheet]=xlsread(blastFile,i);
0054     labels=dataSheet(1,:);
0055     <span class="keyword">if</span> strcmpi(labels{3},<span class="string">'E-value'</span>) &amp;&amp; strcmpi(labels{4},<span class="string">'Alignment length'</span>) &amp;&amp; strcmpi(labels{5},<span class="string">'Identity'</span>)
0056         <span class="comment">%At least one of the organisms must have a model</span>
0057         fromID=find(strcmpi(labels{1},organisms));
0058         toID=find(strcmpi(labels{2},organisms));
0059         <span class="comment">%Check that the organism ids exist and that one of them is the</span>
0060         <span class="comment">%organism of interest</span>
0061         <span class="keyword">if</span> any(fromID) &amp;&amp; any(toID) &amp;&amp; (toID==1 || fromID==1)
0062             <span class="comment">%Check that no gene ids are empty. This could for example be</span>
0063             <span class="comment">%the case if the gene names are wrongly formatted as numbers</span>
0064             <span class="comment">%instead of strings</span>
0065             emptyNames=cellfun(@isempty,dataSheet(2:<span class="keyword">end</span>,1)) | cellfun(@isempty,dataSheet(2:<span class="keyword">end</span>,2));
0066             <span class="keyword">if</span> any(emptyNames)
0067                 <span class="keyword">if</span> all(emptyNames)
0068                     EM=[<span class="string">'Only empty gene names in sheet from '</span> organisms{fromID} <span class="string">' to '</span> organisms{toID}];
0069                     dispEM(EM);
0070                 <span class="keyword">else</span>
0071                     EM=[<span class="string">'Empty gene names in sheet from '</span> organisms{fromID} <span class="string">' to '</span> organisms{toID} <span class="string">'. Ignoring genes with empty names'</span>];
0072                     dispEM(EM,false);
0073                 <span class="keyword">end</span>
0074             <span class="keyword">end</span>
0075             blastStructure(numel(blastStructure)+1).toId=organisms{toID};
0076             blastStructure(numel(blastStructure)).fromId=organisms{fromID};
0077             blastStructure(numel(blastStructure)).fromGenes=dataSheet(2:<span class="keyword">end</span>,1);
0078             blastStructure(numel(blastStructure)).toGenes=dataSheet(2:<span class="keyword">end</span>,2);
0079             blastStructure(numel(blastStructure)).evalue=values(:,1);
0080             blastStructure(numel(blastStructure)).aligLen=values(:,2);
0081             blastStructure(numel(blastStructure)).identity=values(:,3);
0082             
0083             <span class="comment">%Remove matches where any of the values is NaN. This would have</span>
0084             <span class="comment">%been done anyways in getModelFromHomology, but it's neater to</span>
0085             <span class="comment">%do it here</span>
0086             I=isnan(blastStructure(end).evalue) | isnan(blastStructure(end).aligLen) | isnan(blastStructure(end).identity);
0087             blastStructure(end).fromGenes(I)=[];
0088             blastStructure(end).toGenes(I)=[];
0089             blastStructure(end).evalue(I)=[];
0090             blastStructure(end).aligLen(I)=[];
0091             blastStructure(end).identity(I)=[];
0092         <span class="keyword">else</span>
0093             <span class="keyword">if</span> isempty(toID) || isempty(fromID)
0094                 EM=[<span class="string">'The data in sheet '</span> sheets{i} <span class="string">' has no corresponding model. Ignoring sheet'</span>];
0095                 dispEM(EM,false);
0096             <span class="keyword">else</span>
0097                 EM=[<span class="string">'The data in sheet '</span> sheets{i} <span class="string">' does not involve the organism of interest. Ignoring sheet'</span>];
0098                 dispEM(EM,false);
0099             <span class="keyword">end</span>
0100         <span class="keyword">end</span>
0101     <span class="keyword">else</span>
0102         EM=[<span class="string">'The data in sheet '</span> sheets{i} <span class="string">' is not correctly formatted. Ignoring sheet'</span>];
0103         dispEM(EM,false);
0104     <span class="keyword">end</span>
0105 <span class="keyword">end</span>
0106 
0107 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Wed 22-Aug-2018 16:19:49 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>